stylix/mk-target.nix

```diff
@@ -0,0 +1,80 @@
+{
+  name, # name used for creation of options
+  humanName, # name used for description of options
+  autoEnable ? true, # condition for automatically enabling this target
+  extraOptions ? { }, # options to be added under `options.stylix.targets.${name}`
+  configElements ? [ ], # list of functions
+  generalConfig ? null,
+}:
+{ config, lib, ... }:
+let
+  cfg = config.stylix.targets.${name};
+in
+{
+  imports = [
+    {
+      options.stylix.targets.${name}.enable =
+        config.lib.stylix.mkEnableTarget humanName autoEnable;
+    }
+    {
+      options.stylix.targets.${name} = extraOptions;
+    }
+  ];
+
+  config =
+    let
+      functionArgNames =
+        fn:
+        let
+          args = lib.functionArgs fn;
+          invalidArgs = lib.filterAttrs (_: lib.id) args;
+          invalidArgsString = builtins.concatStringsSep ", " (
+            builtins.attrNames invalidArgs
+          );
+        in
+        lib.throwIf (invalidArgs != { })
+          "function arguments should not have default values: ${invalidArgsString}"
+          (builtins.attrNames args);
+      getStylixAttrs =
+        fn:
+        lib.pipe fn [
+          functionArgNames
+          (map (
+            arg:
+            if arg == "cfg" then
+              {
+                name = "cfg";
+                value = cfg;
+              }
+            else if arg == "colors" then
+              {
+                name = arg;
+                value = config.lib.stylix.${arg};
+              }
+            else if config.stylix ? ${arg} then
+              {
+                name = arg;
+                value = config.stylix.${arg};
+              }
+            else
+              throw "stylix: Stylix arg was requested by mk-target but does not exist: ${arg}"
+          ))
+          builtins.listToAttrs
+        ];
```

Comment:
Maybe I'm missing something, but this looks suspiciously similar to `lib.genAttrs`

```nix
getStylixAttrs =
  fn:
  lib.genAttrs functionArgNames (
    arg:
    if arg == "cfg" then
      {
        name = "cfg";
        value = cfg;
      }
    else if arg == "colors" then
    # ...
  );
```

Reply:
Note 998d275218dce69440bf4e3ccf53d4f5715da090 still has the `{ name, value }` attrs needed for `listToAttrs`. These can go with `genAttrs`, since name is always the string value being mapped and value is always the thing returned by the mapping function.

https://noogle.dev/f/lib/genAttrs